### PR TITLE
Community/seasoning and filters

### PR DESCRIPTION
**Pull request by branquinhoaa**
Related issue #594 

## Implements "seasoning": 
 - When an orga member chooses, he can go to '/orga/users/info' and click in reset availability. Then, all the coaches availability will be reset (set to false)

reset

 - In the user profile edit, the coaches will have a checkbox to signalize her "availability" for the upcoming season.

users_edit

 - The coaches' availability will be shown in coaches' profile

user_show

## Enhance the filter/search possibilities
 - filter by availability by location and by the availability 
* We add a column for availability and we want to know what do you think about it

rgsoc teams app 12

### CODE REVIEW

File: app/controllers/community_controller.rb [hunk 1]
```diff
@@ -9,12 +9,17 @@ def index
...
     }
+
+    @countries = User.pluck(:country)
     @users = User.ordered(params[:sort],params[:direction])
         .group('users.id').with_all_associations_joined
     @users = @users.with_assigned_roles if Time.now.utc > (current_season.starts_at || Date.new)
     @users = @users.with_role(params[:role]) if params[:role].present? && params[:role] != 'all'
     @users = @users.with_interest(params[:interest]) if params[:interest].present? && params[:interest] != 'all'
-    @users = Kaminari.paginate_array(@users.search(params[:search])).page(params[:page])
+    @users = @users.as_coach_availability if params[:availability].present?
```
Comment (carpodaster) [inline app/controllers/community_controller.rb]: This @users assignment is getting very long. Could you move this to a private method?

File: config/routes.rb [hunk 2]
```diff
@@ -73,8 +73,10 @@
 
   get 'orga/users/info', to: 'orga/users_info#index', as: :orga_users_info
   patch 'orga/seasons/switch_phase', to: 'orga/seasons#switch_phase', as: :switch_phase
+  post 'orga/community/reset_user_availability', to: 'orga/community#reset_user_availability'
```
Comment (carpodaster) [inline config/routes.rb]: From a RESTful perspective, this would be more of a `PUT` request … I think 🤔

File: app/models/user.rb [hunk 3]
```diff
@@ -87,6 +87,10 @@ def supervisor
       where(name: 'supervisor')
     end
 
+    def coach
```
Comment (carpodaster) [inline app/models/user.rb]: This could be a good time to refactor all these similar methods (except the `admin` one) with some meta programming. Ping me if you need any pointers.

File: app/models/user/availability_switcher.rb [hunk 4]
```diff
@@ -0,0 +1,8 @@
+# frozen_string_literal: true
+
```
Comment (carpodaster) [inline app/models/user/availability_switcher.rb]: I would remove this empty line

File: app/views/users/_roles.html.slim [hunk 5]
```diff
@@ -7,5 +7,9 @@
       label.radio
         = radio_button_tag :role, role == 'all' ? '' : role, params[:role] == role
         = role.capitalize
-
+    h4 Coach availability
+    label.radio
+      = radio_button_tag :availability, true
```
Comment (carpodaster) [inline app/views/users/_roles.html.slim]: I'm not sure about the radio button. 1st, you cannot uncheck it. 2nd, I would imagine that a call-to-action button is better: a button "I am available for the upcoming season" with a direct `PUT` action being triggered (ideally via AJAX so that we don't get a page reload)

File: app/views/users/_roles.html.slim [hunk 6]
```diff
@@ -7,5 +7,9 @@
...
+      | Availability in the upcoming season
```
Comment (carpodaster) [inline app/views/users/_roles.html.slim]: As for the term "upcoming": shall we replace it with the actual year? (`Season#name`)

File: app/controllers/orga/community_controller.rb [hunk 7]
```diff
@@ -0,0 +1,10 @@
+class Orga::CommunityController < Orga::BaseController
+ 
+  def reset_user_availability
+    if User::AvailabilitySwitcher.reset
+      redirect_to orga_users_info_path, flash: { notice: 'Coaches were reset with success.' }
```
Comment (carpodaster) [inline app/controllers/orga/community_controller.rb]: I always try to reduce multiple exits to a single statement at the end of a method/function (unless it's a [guard clause](https://devblast.com/b/what-are-guard-clauses) in its first line). The `redirect_to` acts as a method exit in a controller. You could use the conditional branches to set the `flash[key]` messages and then redirect to `orga_users_info_path` without giving a flash kwarg _after_ the conditional block.

File: app/views/users/_roles.html.slim (hunk 5 above)
Comment (branquinhoaa) [reply app/views/users/_roles.html.slim]: I think you are talking about the checkbox btn that refers to coach availability update. 
This view that you commented above belongs to the filter page.
I removed the checkbox for availability and did an ajax call. I am not sure if that is what you mean...

availability_btn


availability_label

File: app/models/user.rb (hunk 3 above)
Comment (branquinhoaa) [reply app/models/user.rb]: ok, this one we have to study a little bit but we will add the refactor soon. :dancer:

**Comment by branquinhoaa** [comment]
@carpodaster the corrections were made! :+1: Now, @mkalininait  just need to confirm that user#interest will be used instead coach role. Then we can update the code!! :dancer:

File: app/views/users/_roles.html.slim (hunk 5 above)
Comment (carpodaster) [reply app/views/users/_roles.html.slim]: Yes, this is exactly what I meant <3

File: app/models/user.rb (hunk 3 above)
Comment (carpodaster) [reply app/models/user.rb]: 👍

File: app/models/user/availability_switcher.rb [hunk 8]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+class User::AvailabilitySwitcher
+  def self.reset
```
Comment (carpodaster) [inline app/models/user/availability_switcher.rb]: Two thoughts on this class: 

1. I suggest naming a method like this with a bang (`!`) at the end: `reset!` It's a Ruby convention to name "dangerous" method like that (oftentimes methods have a non-bang and a bang version; a "less-dangerous" non-bang version doesn't really apply here but have a look at the stdlib `String` documentation for a couple of good examples on that).
2. Since the name of the class ends in _Switcher_, it is suggested that there is a reverse action (that is: switching back and forth between two states). The sole action in this class only sets all values to `false`, though. It may sound like nit-picking, but naming is very important (if not amongst the most important things): I suggest naming  this class `User::ResetsAvailability` or something along the lines of the. In order to not re-iterate the name of the class in the one public API method (which would look like `User::ResetsAvailability.reset!`) I suggest the common interface for lamba-/proc style callables: `.call` (this time without a bang – confusing, I know :wink:)

File: app/views/community/index.html.slim [hunk 9]
```diff
@@ -23,6 +23,7 @@ table.users.table.table-striped.table-bordered.table-condensed.table-hover
       th = sortable :country
       th = sortable :location, User.human_attribute_name(:location)
       th Time now
+      th Availiability
```
Comment (carpodaster) [inline app/views/community/index.html.slim]: Typo

File: spec/models/user/availability_switcher_spec.rb [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+  describe '.reset' do
+    let!(:coach_available) { create :coach, :available }
+    let!(:user_available) { create :user, :available }
+    before do 
+      User::AvailabilitySwitcher.reset
+    end
+
+    context 'when an user is a coach' do
+      it 'reset the availability to false' do
+        expect(coach_available.reload.availability).to be false
```
Comment (carpodaster) [inline spec/models/user/availability_switcher_spec.rb]: A better way to test a *state change* is to use the `expect { … }.to change { … }` RSpec macro.

Example:

```ruby
expect { User::AvailabilitySwitcher.reset }
  .to change { coach_available.reload }
  .from(true).to(false)
end
```

You don't need the `before` block that way.

File: spec/models/user_spec.rb [hunk 11]
```diff
@@ -63,6 +63,9 @@
 
   describe 'scopes' do
     context 'role scopes' do
+      roles_except_organizer = User::ROLES
```
Comment (carpodaster) [inline spec/models/user_spec.rb]: I'm undecided whether it's good to have the test run use a MP pattern as well. I feel that it would be better for this test to catch regressions by defining the list of roles as a static list, not re-using `User::ROLES`

File: app/assets/javascripts/user_form.js [hunk 12]
```diff
@@ -7,6 +7,29 @@ $(function () {
      $('.form-group.user_company_info').toggle(this.checked);
   }).change(); //ensure visible state matches initially
 
+
+   //update coach availability
+
+   $(function () {
+      var $update_availability = $('[data-behavior="update_availability"]');
+      $update_availability.on('click',function(){
+        var url = "/users/current/update_availability";
```
Comment (michaelem) [inline app/assets/javascripts/user_form.js]: I think hardcoding routes in javascript is somewhat of an issue since it can lead to javascript breaking when those routes changes. This is especially troublesome if these things are not covered by system tests since they then will fail in production. I think the right way to pass the url in this case would probably be to have to url in an additional data field in the element that has the data-behaviour field.

**Comment by mkalininait** [comment]
@carpodaster 
> I'm more and more convinced that just looking at coaches (= users who have been assigned a coach role before) is not the the way to go. This would ignore new users to the teams app that are willing to coach but have never been a coach before. Or RGSoC Alumna. I feel like we need to consider interested_in as well.

I agree! But I'm actually also not sure that we need to know availability of people who have the 'coach' role, but are not currently interested in coaching.

@branquinhoaa I didn't test the PR yet, but I will do after you make this change covering ppl who are *interested in coaching*.

**Comment by branquinhoaa** [comment]
@carpodaster  @mkalininait  About the availability field: it is now available to all user interested_in coaching (not just coaches). 
@carpodaster I removed the AvalabilitySwitcher class because it was useless (now I call a user class method availability_reset!  when I want to reset the user's availability).

**Comment by klappradla** [comment]
Hey @branquinhoaa what's the state of this? Did you already address everything @carpodaster and @mkalininait mentioned and want me to have another look at it 👀  ???

**Comment by branquinhoaa** [comment]
Hey @klappradla! Yeah, I have already finished with the @carpodaster suggestions. 
It would be nice if you can take a look.

**Comment by klappradla** [comment]
Alright, gonna have a look when I'm home later today ✌️

File: app/assets/javascripts/user_form.js [hunk 13]
```diff
@@ -7,6 +7,29 @@ $(function () {
...
+      var $update_availability = $('[data-behavior="update_availability"]');
+      $update_availability.on('click',function(){
+        var url =$(this).data("url");
+        updateAvailability(url);
+      });
+    });
+
+   function updateAvailability(url) {
+      $.ajax({
+        type:     'PUT',
```
Comment (klappradla) [inline app/assets/javascripts/user_form.js]: Also: I think it's the wrong key. It should be 
```js
method: "PUT"
```
since this is about type of the request.

Maybe you were referring to `dataType`??? It's optional here, jQuery will do an intelligent guess, but you could also deliberately set it to whatever you need. Would be
```js
dataType: "script"
```
here I suppose.

File: app/models/community_users.rb [hunk 14]
```diff
@@ -0,0 +1,32 @@
...
+    @users         = users
+    @role          = params[:role].presence
+    @interest      = params[:interest].presence
+    @availability  = params[:availability].presence
+    @search        = params[:search].presence
+    @location      = params[:location].presence
+    @page          = params[:page].presence
+  end
+
+  def all
```
Comment (klappradla) [inline app/models/community_users.rb]: Can you come up with tests for this? You can stub out all the class methods on `User`, I think it would just be helpful to ensure the contract, so that if maybe in the future someone removes one of these class methods, this is caught by our tests.

File: app/models/user.rb [hunk 15]
```diff
@@ -159,9 +155,21 @@ def with_interest(interest)
...
+
+    def as_coach_availability
+      where(availability: true)
+    end
+
     def immutable_attributes
       [:github_handle]
     end
+
+    def reset_availability!
```
Comment (klappradla) [inline app/models/user.rb]: Can you add tests for this? It's a method with rather large impact... Would be good to know if it's really (and only) doing what it should 😉

File: app/controllers/orga/community_controller.rb [hunk 16]
```diff
@@ -0,0 +1,7 @@
+class Orga::CommunityController < Orga::BaseController
+
+  def reset_user_availability
```
Comment (klappradla) [inline app/controllers/orga/community_controller.rb]: Can you add tests for this?

File: app/views/users/_roles.html.slim (hunk 5 above)
[the remark this answers is not in the document]
Comment (branquinhoaa) [reply app/views/users/_roles.html.slim]: In fact it is always unchecked. When you check it, the search starts. I will remove this useless true.

File: app/views/users/edit.html.slim [hunk 17]
```diff
@@ -50,6 +50,16 @@
   = f.input :company_name, label: 'Company name'
   = f.input :company_info, label: 'Company info', hint: 'Where is the office, how much space can you provide, etc.'
   = f.input :bio, label: 'About you', hint: 'You can use basic <a href="http://en.wikipedia.org/wiki/Markdown" target="_blank">Markdown</a> here.'.html_safe
+
+  - if @user.interested_in_coaching?
+    h3.page-header Availability for the next season:
```
[the remark this answers is not in the document]
Comment (branquinhoaa) [reply app/views/users/edit.html.slim]: That is so true!! :joy_cat: I did not think about it...

File: app/controllers/users_controller.rb [hunk 18]
```diff
@@ -49,6 +49,10 @@ def update
     end
   end
 
+  def update_availability
+    current_user.update(availability: true)
```
[the remark this answers is not in the document]
Comment (branquinhoaa) [reply app/controllers/users_controller.rb]: Yeah, that is exactly what I understand by @carpodaster comment. The user can set his availability, but he cannot give up.

[the remark this answers is not in the document]
Comment (klappradla) [reply app/controllers/users_controller.rb]: Alright, perfect! Just wanted to make sure.

**Comment by mkalininait** [comment]
I've tested the PR and have several thoughts.

**User's page (show and edit):**
1) It would be cool if the dependency between "Interested in helping as a coach" and "Availability" on User's page was clearer. E.g. when I set a check that I'm interested to be a coach on the edit page, I see the "I'm available in the next season" checkbox in real time somewhere below. Now this checkbox appears only after I reopen the edit form and it's called "Availability for the next season" which doesn't indicate any relation with the coach status. Same with the title "Availability" on the show page. It's a bit confusing. I think we should at least rename "availability" to "coach availability" everywhere.
2) Apparently I (as a person interested in coaching) can't set my availability status to "Not available" myself. I think it's a necessary feature.
3) Also, once I press the button "I'm available in the next season", I see another button "Successful updated" which I can't press. Why is it a button at all? I'd expect to see a button "I'm NOT available in the next season" or something like that. And if we want to notify that the status is successfully updated, I'd show the message "Successfully updated", not "Successful updated".
4) More about naming: on the edit form the title is "Availability **for** the next season:", but the button and the label are "I am available **in** the next season". At the same time on the show page I see "I am available **for** this season!".  I think we should use "in" (not "for") and "next" (not "this") consistently.

**Resetting :**
1) When I reset coach statuses on the admin panel I see this notification:  "Coaches were reset with success." It's better to say: "Coach availability was reset with success." or something like this.

**Community page:**
1) "Availability in the 2017 season": Why is it a radiobutton? How to reset this filter? What triggers changing the year (say, 2017 to 2018)? I think it should be a checkbox called "Coach availability in the next season". 
2) The country filter doesn't work correctly for me: it shows repeating values.
screen shot 2017-10-03 at 15 20 57

cc @branquinhoaa The summer of Code is kinda over already, so I'm not sure if I should mention you here or not :)

**Comment by branquinhoaasimbiose** [comment]
@mkalininait Of course you should mentioning me here!! I will keep working on Teams app issues :dancer:

**Comment by mkalininait** [comment]
@branquinhoaa Do you want to add more commits or is the PR ready for testing?

**Comment by branquinhoaa** [comment]
@mkalininait Hi! @juuh42dias will fix some things and then I ask you to review ok?

**Comment by klappradla** [comment]
Hey @branquinhoaa 👋 

I know your Summer of Code is over 😿 

But I just wanted to check what's the state of this PR 🕵️ 
Do you plan to back back to it, or should I take it over to add the last bits to it?

**Comment by branquinhoaa** [comment]
@klappradla Hey!! I would like to finish this pr. I will do the last alterations now and then I ask your review, ok? thank you!!!

**Comment by klappradla** [comment]
Awesome 💚 

And yeah, just ping me again.

**Comment by branquinhoaa** [comment]
@klappradla It is ready to the new review!! Thank you

File: app/assets/stylesheets/global.sass [hunk 19]
```diff
@@ -67,6 +67,12 @@ html, body
       &:last-child:after
         content: none
 
+#availability
```
Comment (klappradla) [inline app/assets/stylesheets/global.sass]: This is never used (or at least I cannot find it in this PR 🙈 )

So can you please remove this?

File: app/controllers/community_controller.rb [hunk 20]
```diff
@@ -9,12 +9,12 @@ def index
       mentoring:  'Helping as a Mentor',
       organizing: 'Helping as an Organizer'
     }
+
+    @countries = User.pluck(:country).uniq
     @users = User.ordered(params[:sort], params[:direction])
```
Comment (klappradla) [inline app/controllers/community_controller.rb]: I think it would be a bit more readable if only those things later to be used in the view would be instantiated as instance variables. So how about having:
```ruby
users = User.order...
communtiy = CommunitUsers.new(..., users)
@users = community.all
```

What do you think?

File: app/controllers/users_controller.rb [hunk 21]
```diff
@@ -1,7 +1,7 @@
 class UsersController < ApplicationController
   before_action :set_user, only: [:show, :edit, :update, :destroy, :impersonate]
 
-  load_and_authorize_resource except: [:index, :show, :impersonate, :stop_impersonating]
+  load_and_authorize_resource except: [:index, :show, :impersonate, :stop_impersonating, :set_available]
```
Comment (klappradla) [inline app/controllers/users_controller.rb]: Since this list is getting quite long now, why not "reverse" it and use `only` instead?
```ruby
load_and_authorize_resource only: %i[new create edit update destroy]
```

File: config/routes.rb [hunk 22]
```diff
@@ -73,6 +73,8 @@
 
   get 'orga/users/info', to: 'orga/users_info#index', as: :orga_users_info
   patch 'orga/seasons/switch_phase', to: 'orga/seasons#switch_phase', as: :switch_phase
+  put 'orga/community/reset_user_availability', to: 'orga/community#reset_user_availability'
```
Comment (klappradla) [inline config/routes.rb]: Can you please move this into the `orga` namespaces`?  It will have the `/orga` prefix automatically.

File: config/routes.rb [hunk 23]
```diff
@@ -73,6 +73,8 @@
...
+  put 'users/current/update_availability', to: 'users#set_available'
```
Comment (klappradla) [inline config/routes.rb]: Can you please move this to the `users` resource and make it a member action? It's really hard to find otherwise, given the size of the `routes` file 🙈

File: db/migrate/20170901163452_add_availability_to_users.rb [hunk 24]
```diff
@@ -0,0 +1,5 @@
+class AddAvailabilityToUsers < ActiveRecord::Migration[5.1]
+  def change
+    add_column :users, :availability, :boolean
```
Comment (klappradla) [inline db/migrate/20170901163452_add_availability_to_users.rb]: Hm, personal preference, but I would suggest to **always** put a `null: false` and a default value (`false` in this case) to a boolean value. You'll otherwise end up with a 3 state boolean: `null, true, false` - which can be a potential source of problems later 😉

File: db/migrate/20170905222833_add_default_value_to_availability.rb [hunk 25]
```diff
@@ -0,0 +1,5 @@
+class AddDefaultValueToAvailability < ActiveRecord::Migration[5.1]
+  def change
+    change_column_default :users, :availability, false
```
Comment (klappradla) [inline db/migrate/20170905222833_add_default_value_to_availability.rb]: Why are you using a separate migration for this? I'd suggest to just do it in the migration where the column is added.

File: spec/controllers/orga/community_controller_spec.rb [hunk 26]
```diff
@@ -0,0 +1,17 @@
...
+RSpec.describe Orga::CommunityController do
+  render_views
+
+  context 'with admin logged in' do
+    include_context 'with admin logged in'
+    describe 'PUT reset_user_availability' do
+
+      it 'reset availability for users interested in coaching' do
+        put :reset_user_availability
+        expect(response).to redirect_to (orga_users_info_path)
```
Comment (klappradla) [inline spec/controllers/orga/community_controller_spec.rb]: I'm missing an expectation on the side effects of this request here 👮  My suggestion would be to have something like:
```ruby
expect { put :reset_user_availablity }.to change { user.reload.availablity }.to(false)
```

File: spec/models/community_users.rb [hunk 27]
```diff
@@ -0,0 +1,84 @@
+require 'spec_helper'
+
+describe CommunityUsers do
+  let(:student) { create(:student) }
+  let(:student_looking_for_coach) { create(:student, :interested_in_coaching) }
+  let(:coach) { create(:coach, :available) }
+  describe '#all' do
```
Comment (klappradla) [inline spec/models/community_users.rb]: Can you please add a blank line before the `describe` block? Just easier to read.

File: spec/models/community_users.rb [hunk 28]
```diff
@@ -0,0 +1,84 @@
...
+    before do
+      @params = {
```
Comment (klappradla) [inline spec/models/community_users.rb]: I don't really see the point of switching from `lets` to instance vars here 🙈 

Can you please convert them to `let()` statements?

This also does not need to be in a before block from what I see.

File: spec/controllers/orga/community_controller_spec.rb [hunk 29]
```diff
@@ -0,0 +1,17 @@
+require 'spec_helper'
+
+RSpec.describe Orga::CommunityController do
+  render_views
+
+  context 'with admin logged in' do
+    include_context 'with admin logged in'
+    describe 'PUT reset_user_availability' do
+
+      it 'reset availability for users interested in coaching' do
```
Comment (klappradla) [inline spec/controllers/orga/community_controller_spec.rb]: Typo
> it 'resets the availablity

File: spec/models/community_users.rb [hunk 30]
```diff
@@ -0,0 +1,84 @@
...
+        }
+
+      @users = User.ordered(@params[:sort], @params[:direction])
+                   .group('users.id')
+                   .with_all_associations_joined
+
+      @community = CommunityUsers.new(@params, @users)
+    end
+
+    context 'return filtered users' do
```
Comment (klappradla) [inline spec/models/community_users.rb]: Since there's only one `context` I don't see the point in having it 😉  it just increases the indentation level in this case and does not really give more information.

File: spec/models/community_users.rb [hunk 31]
```diff
@@ -0,0 +1,84 @@
...
+      describe 'with params presence' do
```
Comment (klappradla) [inline spec/models/community_users.rb]: This should be a `context` instead. On a technical level, this does of course make no difference, but things beginning with "with" and "when" are usually `context` blocks and make reading easier.

File: spec/models/community_users.rb [hunk 32]
```diff
@@ -0,0 +1,84 @@
...
+        it 'return user with role specified' do
+          @params[:role] = student.roles.first.name
+
+          community = @community
```
Comment (klappradla) [inline spec/models/community_users.rb]: unnecessary assignment

File: spec/models/community_users.rb [hunk 33]
```diff
@@ -0,0 +1,84 @@
...
+      @users = User.ordered(@params[:sort], @params[:direction])
+                   .group('users.id')
+                   .with_all_associations_joined
+
+      @community = CommunityUsers.new(@params, @users)
+    end
+
+    context 'return filtered users' do
+      describe 'with params presence' do
+        it 'return user with role specified' do
```
Comment (klappradla) [inline spec/models/community_users.rb]: I'd reword to
> it 'returns only users with the specified role'

Same for the following examples.

File: spec/models/community_users.rb [hunk 34]
```diff
@@ -0,0 +1,84 @@
...
+        it 'return user by location' do
+          @params[:location] = student.country
+
+          community = @community
+          expect(community.all).to match_array student
+        end
+      end
+    end
+
+    context 'private methods' do
```
Comment (klappradla) [inline spec/models/community_users.rb]: It's best practice to not test private methods. Unit tests would just cover the public interface and through this also indirectly test the internals.

Could you convert these tests so that you instead test the outcome of these scenarios on the `all` method?

File: spec/models/user_spec.rb [hunk 35]
```diff
@@ -63,6 +63,14 @@
 
   describe 'scopes' do
     context 'role scopes' do
+      roles_except_organizer = [
```
Comment (klappradla) [inline spec/models/user_spec.rb]: I would suggest to couple this closer to the actual implementation. Right now, changing the `ROLES` constant in the users model would probably not even break this test - but it should 😉 

So how about
```ruby
roles-exept_organizer = User::ROLES.except('organizer)
```

File: spec/models/user_spec.rb [hunk 36]
```diff
@@ -126,6 +126,20 @@
         end
       end
 
+      describe '.with_location' do
+        it 'returns users matching one out of many location' do
```
Comment (klappradla) [inline spec/models/user_spec.rb]: Typo: "locations"

File: spec/models/user_spec.rb [hunk 37]
```diff
@@ -391,6 +405,27 @@
     end
   end
 
+  describe '.reset_availability!' do
+    let!(:user) { create(:user, :available) }
+    let!(:user_interested_in_coaching) { create(:user, :available, :interested_in_coaching) }
+
+    before do
+      User.reset_availability!
```
Comment (klappradla) [inline spec/models/user_spec.rb]: I feel like doing this in a before block reduces the possibilities to test the actual behavior of the method.

I would suggest to rather test the "change" and its impact:
```ruby
expect { User.reset_availablity! }.to change { user.reload.availablity }.from(true).to(false)
```

**Comment by F3PiX** [comment]
Very cool feature, however, it doesn't look like this PR is easy to cherrypick-and-live-on. I am going to close this; please reopen if you think it should stay here. 
Maybe we can break down the original issue in smaller steps, and make them available as beginner friendly issues.
